src/cider/nrepl/middleware/util/instrument.clj

```diff
@@ -0,0 +1,445 @@
+                      (or (arglist-reductor
+                           coor [(+ m n) '* (doall (concat head newtail))] specifier)
+                          [(+ m n) nil (doall (concat head newtail))]))
+                    [(+ m n) nil (doall (concat head newtail))]))
+
+            ;; Specifier without modifiers.
+            :else (when-let [[m newtail]
+                             (instrument-forms-per-specifier coor n tail specifier)]
+                    [(+ m n) nil (doall (concat head newtail))]))]
+      (when verbose-debug
+        (print "    Spec" specifier)
+        (if-let [[m] result]
+          (do (when mod (print " with " mod))
+              (print " matched ")
+              (if (> m n)
+                (doall (apply pr (drop n (take m form-list))))
+                (print "nothing.")))
+          (println " didn't match!"))
+        (println))
+      result)))
+
+
+(defn- instrument-args-given-arglist
+  "Try to instrument form assuming its rest follows arglist."
+  [coor args arglist]
+  (when verbose-debug
+    (print "Trying: ")
+    (prn arglist))
+  (let [[n _ result]
+        (reduce (partial arglist-reductor coor) [0 nil args] (reorder-+ arglist))]
+    (when (= n (count args))
+      (when verbose-debug (println "Success!"))
+      result)))
+
+(defn- instrument-special-form-try
+  [coor name args]
+  (some (partial instrument-args-given-arglist coor args)
+        (macro-arglists name)))
+
+(defn- instrument-special-form
+  "Instrument form representing a macro call or special-form."
+  [coor [name & args :as form]]
+  (if-let [result (instrument-special-form-try coor name args)] ;; (first `(defn))
+    (cons name result)
+    form))
+
+;;;; ## Instrumenting Functions and Collections
+;;; This part is quite simple, most of the code is devoted to checking
+;;; form-types and special cases. The idea here is that we walk
+;;; through collections and function arguments looking for interesting
+;;; things around which we'll wrap a breakpoint. Interesting things
+;;; are most function-forms and vars satisfying `interesting-symbol?`.
+(defn- instrument-coll
+  "Instrument a general collection."
+  [coor coll]
+  (let [out (map #(instrument (conj coor %1) %2)
+                 (range (count coll))
+                 coll)]
+    (if (listy? coll)
+      out
+      (into (empty coll) out))))
+
+(defn- instrument-function-like-form
+  "Instrument form representing a function/macro call or special-form."
+  [coor [name & args :as form]]
+  (if (symbol? name)
+    (let [name (or (ns-resolve *ns* name) name)]
+      (if (or (resolve-special name)
+              (:macro (meta name)))
+        (if (irrelevant-return-value-macros name)
+          (instrument-special-form coor form)
+          `(breakpoint ~(instrument-special-form coor form) '~coor))
+        `(breakpoint ~(instrument-coll coor form) '~coor)))
+    `(breakpoint ~(instrument-coll coor form) '~coor)))
+
+(defn instrument
```

Comment:
Indeed. The `coor` vector is passed around through all `instrument-*` commands and always makes it to the breakpoints, so we can store any information we want in it and it will be given to the breakpoints. A series of integers is appended to this vector, but anything initially contained in it is left untouched.